compiler/src/dotty/tools/dotc/interpreter/Interpreter.scala

```diff
@@ -0,0 +1,129 @@
+ *
+ *  The interpreter assumes that all calls in the trees are to code that was
+ *  previously compiled and is present in the classpath of the current context.
+ */
+class Interpreter(implicit ctx: Context) {
+  import tpd._
+
+  private[this] val classLoader = {
+    val urls = ctx.settings.classpath.value.split(':').map(cp => java.nio.file.Paths.get(cp).toUri.toURL)
+    new URLClassLoader(urls, getClass.getClassLoader)
+  }
+
+  /** Returns the interpreted result of interpreting the code represented by the tree.
+   *  Return Some of the result or None if some error happen during the interpretation.
+   */
+  def interpretTree[T](tree: Tree)(implicit ct: ClassTag[T]): Option[T] = {
+    try {
+      interpretTreeImpl(tree) match {
+        case obj: T => Some(obj)
+        case obj =>
+          ctx.error(s"Interpreted tree returned a result of an unexpected type. Expected ${ct.runtimeClass} but was ${obj.getClass}", tree.pos)
+          throw new StopInterpretation
+      }
+    } catch {
+      case _: StopInterpretation => None
+    }
+  }
+
+  /** Returns the interpreted result of interpreting the code represented by the tree.
+   *  Returns the result of the interpreted tree.
+   *
+   *  If some error is encountered while interpreting a ctx.error is emited and a StopInterpretation is thrown.
+   */
+  private def interpretTreeImpl(tree: Tree): Object = {
+    try {
+      tree match {
+        case Apply(_, quote :: Nil) if tree.symbol eq defn.quoteMethod =>
+          new RawExpr(quote)
+        case TypeApply(_, quote :: Nil) if tree.symbol eq defn.typeQuoteMethod =>
+          new RawType(quote)
+
+        case Literal(Constant(c)) =>
+          c.asInstanceOf[AnyRef]
+
+        case Apply(fn, args) if fn.symbol.isConstructor =>
+          val cls = fn.symbol.owner
+          val clazz = classLoader.loadClass(cls.symbol.fullName.toString)
+          val paramClasses = paramsSig(fn.symbol)
+          val args1: List[Object] = args.map(arg => interpretTreeImpl(arg))
+          clazz.getConstructor(paramClasses: _*).newInstance(args1: _*).asInstanceOf[Object]
+
+        case Apply(fun, args) if fun.symbol.isStatic =>
+          val clazz = classLoader.loadClass(fun.symbol.owner.companionModule.fullName.toString)
+          val paramClasses = paramsSig(fun.symbol)
+          val args1: List[Object] = args.map(arg => interpretTreeImpl(arg))
+          val method = clazz.getMethod(fun.symbol.name.toString, paramClasses: _*)
+          method.invoke(null, args1: _*)
+
+        case tree: RefTree if tree.symbol.isStatic =>
+          val clazz = classLoader.loadClass(tree.symbol.owner.companionModule.fullName.toString)
+          val method = clazz.getMethod(tree.name.toString)
+          method.invoke(null)
+
+        case tree: RefTree if tree.symbol.is(Module) =>
+          ??? // TODO
+
+        case Inlined(_, bindings, expansion) =>
+          if (bindings.nonEmpty) ??? // TODO evaluate bindings and add environment
```

Comment:
Once we address this TODO we could also interpret blocks; it's the same problem.

Reply:
Furthermore, we could also call non-static methods on any object available in the environment.